scylla-cdc-printer/src/printer.rs

```diff
@@ -0,0 +1,294 @@
+
+pub struct CDCLogPrinter {
+    // Tells the worker to stop
+    // Usage of the "watch" channel will make it possible to change the the timestamp later,
+    // for example if somebody loses patience and wants to stop now not later
+    end_timestamp: tokio::sync::watch::Sender<Duration>,
+}
+
+impl CDCLogPrinter {
+    // Creates a printer and runs it immediately
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        session: &Arc<Session>,
+        keyspace: String,
+        table_name: String,
+        start_timestamp: Duration,
+        end_timestamp: Duration,
+        window_size: Duration,
+        safety_interval: Duration,
+        sleep_interval: time::Duration,
+    ) -> (Self, RemoteHandle<Result<anyhow::Result<()>, JoinError>>) {
+        let (end_timestamp_sender, end_timestamp_receiver) =
+            tokio::sync::watch::channel(end_timestamp);
+        let mut worker = CDCLogPrinterWorker::new(
+            session,
+            keyspace,
+            table_name,
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            end_timestamp_receiver,
+        );
+        let (_, handle) = tokio::task::spawn(async move { worker.run().await }).remote_handle();
+        let printer = CDCLogPrinter {
+            end_timestamp: end_timestamp_sender,
+        };
+        (printer, handle)
+    }
+
+    // Tell the worker to set the end timestamp and then stop
+    pub fn stop_at(&mut self, when: Duration) {
+        let _ = self.end_timestamp.send(when);
+    }
+
+    // Tell the worker to stop immediately
+    pub fn stop(&mut self) {
+        self.stop_at(Duration::min_value());
+    }
+}
+
+struct CDCLogPrinterWorker {
+    session: Arc<Session>,
+    keyspace: String,
+    table_name: String,
+    start_timestamp: Duration,
+    end_timestamp: Duration,
+    sleep_interval: time::Duration,
+    window_size: Duration,
+    safety_interval: Duration,
+    readers: Vec<Arc<StreamReader>>,
+    end_timestamp_receiver: tokio::sync::watch::Receiver<Duration>,
+}
+
+impl CDCLogPrinterWorker {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        session: &Arc<Session>,
+        keyspace: String,
+        table_name: String,
+        start_timestamp: Duration,
+        end_timestamp: Duration,
+        window_size: Duration,
+        safety_interval: Duration,
+        sleep_interval: time::Duration,
+        end_timestamp_receiver: tokio::sync::watch::Receiver<Duration>,
+    ) -> CDCLogPrinterWorker {
+        CDCLogPrinterWorker {
+            session: Arc::clone(session),
+            keyspace,
+            table_name,
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            readers: vec![],
+            end_timestamp_receiver,
+        }
+    }
+
+    pub async fn run(&mut self) -> anyhow::Result<()> {
```

Comment:
This version of `CDCLogPrinterWorker::run` is a big step in a good direction. The event loop pattern is a good idea, but here are still some issues related to how generation change is handled and when the function is supposed to exit. Nevertheless, I think we are close to the finish line.

To help you out, here is how I think a correct loop should look like. Given this additional state:

```rust
let next_generation: Option<Generation> = None;
let had_first_generation: bool = false;
let err: Option<anyhow::Error> = None;
```

...the loop iteration should look like this:

- Wait for an event
	- When a task has finished:
		- If there was an error, store this error in `err` for later
	- When a new generation appears (wait for this event only if `next_generation` is empty - you can put an `if` in the `select!` macro, see the [docs](https://docs.rs/tokio/latest/tokio/macro.select.html)):
		- Store the generation in `next_generation` 
		- Set `had_first_generation` to true
		- Update the upper timestamps of `self.readers`
	- When the end timestamp is updated:
		- Set the end timestamp field
		- Update the upper timestamps of `self.readers`

- If there are no tasks running:
	- If there is `err`:
		- Return the error
	- If there is `next_generation`:
		- If the `next_generation` is larger than the end timestamp:
			- Return success
		- Else:
			- Spawn new tasks
			- Clear `next_generation`
	- Else if `had_first_generation`:
		- Return success

Reply:
Do we still want to stop all the tasks if one of them has failed (`set_upper_timestamp(chrono::Duration::min_value())`) or do we abandon this idea?

Reply:
Thank you for detailed break down. 

Why are we returning when there is no `next_generation` and `had_first_generation` is `true`? What makes us sure that there is no `next_generation` coming in the next iteration? I do not get it. 

More generally: why are we handling `first_generation` differently?

Reply:
> Do we still want to stop all the tasks if one of them has failed (`set_upper_timestamp(chrono::Duration::min_value())`) or do we abandon this idea?

Yes, we do. Sorry, I missed it in the break down - I updated it just now.

Reply:
> Thank you for detailed break down.
> 
> Why are we returning when there is no `next_generation` and `had_first_generation` is `true`? What makes us sure that there is no `next_generation` coming in the next iteration? I do not get it.

I see there is a potentially problematic situation:

1. We are reading generation `G_1`,
2. The printer is told to stop at timestamp `T`,
3. A generation `G_2` appears which is less than `T`,
4. Due to some issues fetching generations doesn't succeed for some time and the printer doesn't learn about `G_2`,
5. The printer reaches timestamp `T` in generation `G_1`,
6. The printer decides to stop, despite not reading data from range `G_2..T` in `G_2`.

I talked with @kbr- and he suggested triggering a generation fetch and making sure it succeeds when we reach point (5), so that we know if there are more generations to process. Generations are advertised 1 minute before, so the assumption is that a successful fetch will inform us about the problematic generation.

However, not to block the progress of this PR further, I think we can leave it as a FIXME in the code. Please include a short description of the problem and the solution/link to this GitHub comment.

> 
> More generally: why are we handling `first_generation` differently?

The `had_first_generation` flag was supposed to differentiate two cases: 1) we are waiting for the first generation 2) we reached the end timestamp and there was no new generation. It was supposed to prevent a situation where the printer quits due to the end timestamp being set before the first generation arrives.

Perhaps we can get rid of it if we wait for the first generation separately (despite me telling you otherwise in the previous round of review). Waiting could be aborted in case a minimum timestamp is set as the end timestamp.